### PR TITLE
Add wav2vec2 loss function in self_supervised_learning training recipe

### PR DESCRIPTION
**Pull request by Zooll**
Add wav2vec2 loss function in the self_supervised_learning training recipe to support Wav2Vec2 pre-training.

### CODE REVIEW

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+from typing import Optional, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+
+
+def compute_contrastive_loss(x, mask_indices, targets, neg_is_pos, reduce, logit_temp = 0.1):
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:8]: could you also add docstring to the function?
```suggestion
def compute_contrastive_loss(x: torch.Tensor, mask_indices: torch.Tensor, targets: torch.Tensor, neg_is_pos: torch.Tensor, reduce: str, logit_temp: int = 0.1):
```

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        x[mask_indices]
+        .view(x.size(0), -1, x.size(-1))
+        .unsqueeze(0)
+        .expand(targets.shape)
+    )
+    logits = torch.cosine_similarity(x.float(), targets.float(), dim=-1).float()
+    logits /= logit_temp
+    if neg_is_pos.any():
+        logits[1:][neg_is_pos] = float("-inf")
+    target = logits.new_zeros(logits.size(1) * logits.size(2), dtype=torch.long)
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:19]:
```suggestion
    target = logits.new_zeros(logits.size(1) * logits.size(2), dtype=torch.long, device=logits.device)
```

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    logits = logits.transpose(0, 2)
+    logits = logits.reshape(-1, logits.size(-1))
+    loss = F.cross_entropy(
+        logits,
+        target,
+        reduction="sum" if reduce else "none",
+    )
+    sample_size = target.numel()
+    return loss, sample_size, logits
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:28]: I think we can remove the logits from output, since it is not used in the loss function
```suggestion
    return loss, sample_size
```

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    Returns:
+        (Tensor, float)
+        Tensor: The desired loss Tensor.
+        float: Sample size according to mask_indices
+    """
+    assert y is not None
+    assert mask_indices is not None
+    assert mask_indices.sum() == y.shape[0] * y.shape[1]
+
+    # 4. compute targets
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:55]: the number can be removed

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    neg_is_pos = (y == negatives).all(-1)
+    y = y.unsqueeze(0)
+    targets = torch.cat([y, negatives], dim=0)
+
+    # 5. compute losses
+    loss, sample_size, _ = compute_contrastive_loss(
+        x, mask_indices, targets, neg_is_pos, reduce
+    )
+    loss = loss.float()
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:64]:
```suggestion
```

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        target,
+        reduction="sum" if reduce else "none",
+    )
+    sample_size = target.numel()
+    return loss, sample_size, logits
+
+def wav2vec2_loss(
+    x: Tensor,
+    mask_indices: Tensor,
+    y: Tensor,
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:33]: we can rename `y` to `postive` to represent its meaning.

**Review by nateanl — COMMENTED** [review]
Added some comments, could you address these? Otherwise LGTM!

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    neg_is_pos: Tensor,
+    reduce: bool,
+    logit_temp: float = 0.1):
+    """
+    Computes the contrastive loss used in Wav2Vec2 loss function.
+
+    Args:
+        x (Tensor): Input embeddings of shape (batch_size, sequence_length, hidden_size)
+        mask_indices (Tensor): Indices to mask negative samples
+        targets (Tensor): Labels indicating positive samples of shape (batch_size, sequence_length, hidden_size).
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:21]: The shape of target should be `(num_negative + 1, batch, sequence_length, hidden_size)`, right?

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        neg_is_pos (Tensor): Boolean tensor indicating whether negative samples should be treated as positives
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:22]: could you also add the shape of `neg_is_pos`?  it is supposed to be `(batch, sequence_length)`.

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    )
+    sample_size = target.numel()
+    return loss, sample_size
+
+def wav2vec2_loss(
+    x: Tensor,
+    mask_indices: Tensor,
+    positives: Tensor,
+    negatives: Tensor, 
+    reduce: Optional[bool]=True
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:56]: we can simply the `reduce` argument, for example, rename it to `reduction`  with `str` type and only accept `"sum"` and `"none"`.

File: examples/self_supervised_learning/losses/_wav2vec2_loss.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+) -> Tuple[Tensor, float]:
+    """Compute Wav2Vec2 loss.
+
+    Args:
+        x (Tensor): The masked sequences of probability distribution.
+        mask_indices (Tensor): The mask indices.
+        positives (Tensor): The positives, prior to negative sampling.
+        negatives (Tensor): The negative samples.
```
Comment (nateanl) [inline examples/self_supervised_learning/losses/_wav2vec2_loss.py:64]: could you add shapes for `positives` and `negatives`?

**Review by nateanl — APPROVED** [review]
just some nits and LGTM!

**Review by mthrok — APPROVED** [review]
